### PR TITLE
Implement a generic Destination Propagation optimization on MIR

### PR DESCRIPTION
**Pull request by jonas-schievink**
This takes the work that was originally started by @eddyb in https://github.com/rust-lang/rust/pull/47954, and then explored by me in https://github.com/rust-lang/rust/pull/71003, and implements it in a general (ie. not limited to acyclic CFGs) and dataflow-driven way (so that no additional infrastructure in rustc is needed).

The pass is configured to run at `mir-opt-level=2` and higher only. To enable it by default, some followup work on it is still needed:
* Performance needs to be evaluated. I did some light optimization work and tested against `tuple-stress`, which caused trouble in my last attempt, but didn't go much in depth here.
  * We can also enable the pass only at `opt-level=2` and higher, if it is too slow to run in debug mode, but fine when optimizations run anyways.
* Debuginfo needs to be fixed after locals are merged. I did not look into what is required for this.
* Live ranges of locals (aka `StorageLive` and `StorageDead`) are currently deleted. We either need to decide that this is fine, or if not, merge the variable's live ranges (or remove these statements entirely – https://github.com/rust-lang/rust/issues/68622).

Some benchmarks of the pass were done in https://github.com/rust-lang/rust/pull/72635.

### CODE REVIEW

**Comment by jonas-schievink** [comment]
Urg, there's still a miscompilation in there somewhere.

**Comment by jrmuizel** [comment]
#45663, #58082, #56172 are some unnecessary memcpy issues. It would be interesting to know how many of them are solved by this approach.

**Comment by goffrie** [comment]
> #45663, #58082, #56172 are some unnecessary memcpy issues. It would be interesting to know how many of them are solved by this approach.

I tried it out and it looks like #56172 is improved with this PR (no memcpys) but the others are unchanged.

**Comment by dotdash** [comment]
This modified example from #56172 regresses with this pass:

```rust
#[inline(never)]
pub fn g(clip: Option<&bool>) {
    clip.unwrap();
    let item = SpecificDisplayItem::PopStackingContext;
    do_item(&DI {
            item,
    });
    do_item(&DI {
            item,
    });
}

pub enum SpecificDisplayItem {
    PopStackingContext,
    Other([f64; 22]),
}

struct DI {
    item: SpecificDisplayItem,
}


fn do_item(di: &DI) { unsafe { ext(di) } }
extern {
    fn ext(di: &DI);
}
```

Nightly optimizes this properly, but with this pass, one of the `DI` instances is directly initialized (good), but then this is used to initialized the other instance, which is bad because that means that the memcpy can no longer be omitted and the lifetimes of those stack allocations are now overlapping so stack coloring can no longer collapse them into a single stack slot, increasing memory usage.

**Comment by dotdash** [comment]
Since I always expected this to work the other way around, replacing dest with source, not the other way around, what's the reasoning behind doing it this way? Feel free to just point me somewhere where I can read up on it myself, I just didn't find anything right away.

**Comment by bjorn3** [comment]
The return variable is always called `_0`, so if you have `_0 = _10`, then every assignment of `_10` must be replaced with an assignment of `_0` to perform destination propagation.

**Comment by dotdash** [comment]
That's what RVO does, assigning the return slot to a value that would otherwise use a separate stack slot, but that's just one way. Doing copy propagation works just as well as long as there is only one possible value that is being assigned.

```
_5 = "Hello"
_0 = _5
```
is transformed to the following by this pass:
```
_0 = "Hello"
nop
```
but copy propagation would lead to:
```
_5 = "Hello"
_0 = "Hello"
```
Something like dead store elimination would then have clean up the dead assignment to `_5` (or have a built-in DSE).

The example from above is akin to
```
_5 = "Hello"
_2 = _5
call(_2)
_3 = _5
call(_3)
```
and gets converted to
```
_3 = "Hello"
_2 = _3
call(_2)
nop
call(_3)
```
while copy propagation would give
```
_5 = "Hello"
_2 = "Hello"
call(_2)
_3 = "Hello"
call(_3)
```

The difference being that copy propagation doesn't force the two destinations to be alive at the same time, and apparently allowing more optimizations to happen e.g. in case of aggregates.

The case where the approach taken here works better is like this:
```
if _10, goto bb1 else goto bb2

bb1:
  _2 = "Hello"
  goto bb3

bb2:
  _2 = "Bye"
  goto bb3

bb3:
  _5 = _2
```

I'm wondering whether it's more common to use a single value in multiple places, or assign to single destination from a value that has been set in multiple places.

I suppose the truth is somewhere in between... Maybe limiting this pass to cases where only a single replacement is possible is useful?

**Comment by jonas-schievink** [comment]
Rebased.

r? @oli-obk

File: compiler/rustc_mir/src/transform/dest_prop.rs [hunk 1]
```diff
@@ -0,0 +1,1056 @@
...
+//!   It must also not contain any indexing projections, since those take an arbitrary `Local` as
+//!   the index, and that local might only be initialized shortly before `dest` is used.
+//!
+//!   Subtle case: If `dest` is a, or projects through a union, then we have to make sure that there
+//!   remains an assignment to it, since that sets the "active field" of the union. But if `src` is
+//!   a ZST, it might not be initialized, so there might not be any use of it before the assignment,
+//!   and performing the optimization would simply delete the assignment, leaving `dest`
+//!   uninitialized.
+//!
+//! * `src` must be a bare `Local` without any indirections or field projections (FIXME: Why?).
```
Comment (oli-obk) [inline compiler/rustc_mir/src/transform/dest_prop.rs:47]: this is not true anymore as far as I can tell

File: compiler/rustc_mir/src/transform/dest_prop.rs [hunk 2]
```diff
@@ -0,0 +1,1056 @@
...
+                    .collect();
+                let projection = tcx.intern_place_elems(&projection);
+
+                // Replace with the final `Place`.
+                self.map[local] = Some(Place { local: base, projection });
+            }
+        }
+    }
+
+    fn for_src(&self, src: Local) -> Option<&Place<'tcx>> {
```
Comment (oli-obk) [inline compiler/rustc_mir/src/transform/dest_prop.rs:288]: places are `Copy` nowadays

File: compiler/rustc_mir/src/transform/dest_prop.rs (hunk 1 above)
Comment (oli-obk) [reply compiler/rustc_mir/src/transform/dest_prop.rs:47]: Nevermind it's still checked really early, but the code looks to me like it handles projection merging

**Comment by oli-obk** [comment]
I can't hold the entire thing in my head at a time, but the parts all do what they describe and the overarching logic also makes sense.

r=me with the style nit resolved and a rebase

once merged we can experiment with removing NRVO and copy prop, setting this to mir-opt-level 1 and checking perf

File: compiler/rustc_mir/src/transform/dest_prop.rs (hunk 1 above)
Comment (jonas-schievink) [reply compiler/rustc_mir/src/transform/dest_prop.rs:47]: It handles projections in `dest`, but not in `src`

**Comment by jonas-schievink** [comment]
> Assertion failed: ((((rwlock_t *)*rwl)->valid == LIFE_RWLOCK) && (((rwlock_t *)*rwl)->busy > 0)), file ../../src/mingw-w64/mingw-w64-libraries/winpthreads/src/rwlock.c, line 40

Scary but spurious?

@bors retry

**Comment by mati865** [comment]
@jonas-schievink is there issue open for that assertion?

**Comment by jonas-schievink** [comment]
No, I haven't seen that happen anywhere else so far

File: compiler/rustc_mir/src/transform/dest_prop.rs [hunk 3]
```diff
@@ -0,0 +1,1057 @@
...
+            }
+            TerminatorKind::InlineAsm {
+                template: _,
+                operands,
+                options: _,
+                line_spans: _,
+                destination: _,
+            } => {
+                // The intended semantics here aren't documented, we just assume that nothing that
+                // could be written to by the assembly may overlap with any other operands.
```
Comment (Amanieu) [inline compiler/rustc_mir/src/transform/dest_prop.rs:660]: The semantics of inline assembly are documented in the unstable book:
> Operand expressions are evaluated from left to right, just like function call arguments. After the `asm!` has executed, outputs are written to in left to right order. This is significant if two outputs point to the same place: that place will contain the value of the rightmost output.

**Pull request merged**